### PR TITLE
feat: update mermaid rendering flow

### PR DESCRIPTION
**Pull request by antran22**
Split from #5850, this PR only concerns with addressing #4802.

Performance improvement:
- Each diagram now maintains its rendering state.
- Updating code of one diagram only re-render its output. Before we re-render the whole doc page on each change. This is done by each diagram maintaining its current text input, and only update when the text input changes.
- Add debounce for updating the diagram.

Usability improvement:
- Display MermaidJS render error to the user, instead of just "Error rendering Diagram"

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -1102,10 +1102,10 @@
   resolved "https://registry.yarnpkg.com/@benrbray/prosemirror-math/-/prosemirror-math-0.2.2.tgz#5b2aeb7c8c5a41d2e39317fe1fbd7040fd347cc9"
   integrity sha512-n+V8MNKaQ9HtA1IASzoBFwthFY55kpu2I+0aF103AbqUw5eM8YlxHeltnLqjnYRVY4/a6A9t9YlBMBQOli5jgw==
 
-"@braintree/sanitize-url@^6.0.0":
-  version "6.0.2"
-  resolved "https://registry.yarnpkg.com/@braintree/sanitize-url/-/sanitize-url-6.0.2.tgz#6110f918d273fe2af8ea1c4398a88774bb9fc12f"
-  integrity sha512-Tbsj02wXCbqGmzdnXNk0SOF19ChhRU70BsroIi4Pm6Ehp56in6vch94mfbdQ17DozxkL3BAVjbZ4Qc1a0HFRAg==
+"@braintree/sanitize-url@^6.0.0", "@braintree/sanitize-url@^6.0.1":
```
Comment (tommoor) [inline yarn.lock:1105]: Can we remove the `yarn.lock` changes?

File: shared/editor/extensions/Mermaid.ts [hunk 2]
```diff
@@ -13,8 +16,84 @@ import { findBlockNodes } from "../queries/findChildren";
...
+        element
+      );
+    } catch (error) {
+      const isEmpty = block.node.textContent.trim().length === 0;
+
+      if (isEmpty) {
+        element.innerText = "Empty diagram";
+        element.classList.add("empty");
+      } else {
+        element.innerText = `Error rendering diagram\n ${error}`;
```
Comment (tommoor) [inline shared/editor/extensions/Mermaid.ts:73]:
```suggestion
        element.innerText = `Error rendering diagram\n\n${error}`;
```

Should look neater and avoid that hanging space.

image

File: shared/editor/extensions/Mermaid.ts [hunk 3]
```diff
@@ -13,8 +16,84 @@ import { findBlockNodes } from "../queries/findChildren";
...
 
+type RendererFunc = (block: { node: Node; pos: number }) => void;
+
+class MermaidRenderer {
+  readonly diagramId: string;
+  element?: HTMLElement;
+  readonly elementId: string;
+  private currentTextContent = "";
+  private _rendererFunc?: RendererFunc;
+  constructor() {
```
Comment (tommoor) [inline shared/editor/extensions/Mermaid.ts:30]:
```suggestion

  constructor() {
```

File: shared/editor/extensions/Mermaid.ts [hunk 4]
```diff
@@ -23,74 +102,64 @@ function getNewState({
...
+
+    initialized = true;
+  }
+
   blocks.forEach((block) => {
     const existingDecorations = pluginState.decorationSet.find(
       block.pos,
-      block.pos + block.node.nodeSize
+      block.pos + block.node.nodeSize,
+      (spec) => !!spec["diagramId"]
```
Comment (tommoor) [inline shared/editor/extensions/Mermaid.ts:136]:
```suggestion
      (spec) => !!spec.diagramId
```

File: shared/editor/extensions/Mermaid.ts [hunk 5]
```diff
@@ -23,74 +102,64 @@ function getNewState({
...
+          decoration.from,
+          decoration.to,
+          block.pos,
+          block.pos + block.node.nodeSize
+        )
+      )
+    );
+
+    const renderer: MermaidRenderer =
+      bestDecoration?.spec?.["renderer"] ?? new MermaidRenderer();
```
Comment (tommoor) [inline shared/editor/extensions/Mermaid.ts:151]:
```suggestion
      bestDecoration?.spec?.renderer ?? new MermaidRenderer();
```

File: shared/editor/extensions/Mermaid.ts [hunk 6]
```diff
@@ -23,74 +102,64 @@ function getNewState({
...
       block.pos,
-      block.pos + block.node.nodeSize
+      block.pos + block.node.nodeSize,
+      (spec) => !!spec["diagramId"]
     );
 
-    // Attempt to find the existing diagramId from the decoration, or assign
-    // a new one if none exists yet.
-    const diagramId = existingDecorations[0]?.spec["diagramId"] ?? uuidv4();
+    const bestDecoration = last(
```
Comment (tommoor) [inline shared/editor/extensions/Mermaid.ts:139]: Not sure I follow what's happening here, either we have an existing renderer/decoration with the correct ID – or not, no? Be worth a code comment.

File: shared/editor/extensions/Mermaid.ts [hunk 7]
```diff
@@ -13,8 +16,84 @@ import { findBlockNodes } from "../queries/findChildren";
...
+    this.element =
+      document.getElementById(this.elementId) || document.createElement("div");
+    this.element.id = this.elementId;
+    this.element.classList.add("mermaid-diagram-wrapper");
+    return this.element;
+  }
+
+  async renderImmediately(block: { node: Node; pos: number }) {
+    const diagramId = this.diagramId;
+    const element = this.initializeElement();
```
Comment (tommoor) [inline shared/editor/extensions/Mermaid.ts:48]: Move to constructor?

File: shared/editor/extensions/Mermaid.ts [hunk 8]
```diff
@@ -13,8 +16,84 @@ import { findBlockNodes } from "../queries/findChildren";
...
+        element.classList.add("parse-error");
+      }
+    }
+  }
+
+  get render(): RendererFunc {
+    if (this._rendererFunc) {
+      return this._rendererFunc;
+    }
+    this._rendererFunc = debounce<RendererFunc>(this.renderImmediately, 1000);
```
Comment (tommoor) [inline shared/editor/extensions/Mermaid.ts:83]:
```suggestion
    this._rendererFunc = debounce<RendererFunc>(this.renderImmediately, 500);
```

1s is a bit too laggy here

**Review by tommoor — COMMENTED** [review]
This is great, I'm confident we can get this merged soon

File: shared/editor/extensions/Mermaid.ts (hunk 8 above)
Comment (antran22) [reply shared/editor/extensions/Mermaid.ts:83]: Agree

File: shared/editor/extensions/Mermaid.ts (hunk 6 above)
Comment (antran22) [reply shared/editor/extensions/Mermaid.ts:139]: This issue happens when you have one diagram right under another. For example:
- Block 1: Span from 101 to 200
- Block 2: Span from 201 to 300

If you delete some text in Block 1, at this point the `block` value may span both Block 1 and 2, and we may accidentally get the Decoration for Block 2 when we update Block 1.

I have not fully investigated this behavior, however I found out that I can build a heuristic by finding the old Decoration node with most overlap over the recently edited block.

File: yarn.lock (hunk 1 above)
Comment (antran22) [reply yarn.lock:1105]: You mean by un-committing this whole file?

Comment (tommoor) [reply yarn.lock:1105]: Yep

**Review by tommoor — APPROVED** [review]

**Pull request merged**